code/game/machinery/computer/sm_monitor.dm

```diff
@@ -0,0 +1,142 @@
+	var/list/data = list()
+
+	if(istype(active))
+		var/turf/T = get_turf(active)
+		// If we somehow delam during this proc, handle it somewhat
+		if(!T)
+			active = null
+			refresh()
+			return
+		var/datum/gas_mixture/air = T.return_air()
+		if(!air)
+			active = null
+			return
+
+		data["active"] = TRUE
+		data["SM_integrity"] = active.get_integrity()
+		data["SM_power"] = active.power
+		data["SM_ambienttemp"] = air.temperature
+		data["SM_ambientpressure"] = air.return_pressure()
+		//data["SM_EPR"] = round((air.total_moles / air.group_multiplier) / 23.1, 0.01)
+		var/other_moles = air.total_trace_moles()
+		var/TM = air.total_moles()
+		if(TM)
+			data["SM_gas_O2"] = round(100*air.oxygen/TM,0.01)
+			data["SM_gas_CO2"] = round(100*air.carbon_dioxide/TM,0.01)
+			data["SM_gas_N2"] = round(100*air.nitrogen/TM,0.01)
+			data["SM_gas_PL"] = round(100*air.toxins/TM,0.01)
+			if(other_moles)
+				data["SM_gas_OTHER"] = round(100 * other_moles / TM, 0.01)
+			else
+				data["SM_gas_OTHER"] = 0
+		else
+			data["SM_gas_O2"] = 0
+			data["SM_gas_CO2"] = 0
+			data["SM_gas_N2"] = 0
+			data["SM_gas_PH"] = 0
+			data["SM_gas_OTHER"] = 0
+	else
+		var/list/SMS = list()
+		for(var/obj/machinery/power/supermatter_shard/S in supermatters)
+			var/area/A = get_area(S)
+			if(!A)
+				continue
+
+			SMS.Add(list(list(
+			"area_name" = A.name,
+			"integrity" = S.get_integrity(),
+			"uid" = S.UID()
+			)))
+
+		data["active"] = FALSE
+		data["supermatters"] = SMS
+
+	return data
+
+/**
+  * Supermatter List Refresher
+  *
+  * This proc loops through the list of supermatters in the atmos SS and adds them to this console's cache list
+  */
+/obj/machinery/computer/sm_monitor/proc/refresh()
+	supermatters = list()
+	var/turf/T = get_turf(tgui_host()) // Get the TGUI host incase this ever turned into a supermatter monitoring module for AIs to use or something
+	if(!T)
+		return
+	for(var/obj/machinery/power/supermatter_shard/S in SSair.atmos_machinery)
+		// Delaminating, not within coverage, not on a tile.
+		if(!(is_station_level(S.z) || is_mining_level(S.z)  || atoms_share_level(S, T) || !istype(S.loc, /turf/simulated/)))
+			continue
+		supermatters.Add(S)
+
+	if(!(active in supermatters))
+		active = null
+
+/obj/machinery/computer/sm_monitor/process()
+	if(stat & (NOPOWER|BROKEN))
+		return FALSE
+
+	if(active)
+		var/new_status = active.get_status()
+		if(last_status != new_status)
+			last_status = new_status
+			if(last_status == SUPERMATTER_ERROR)
+				last_status = SUPERMATTER_INACTIVE
+			icon_screen = "smmon_[last_status]"
+			update_icon()
+
+	return TRUE
+
+/obj/machinery/computer/sm_monitor/tgui_act(action, params)
+	if(..())
+		return
+
+	. = TRUE
+
+	switch(action)
+		if("refresh")
+			refresh()
+
+		if("view")
+			var/newuid = params["view"]
+			for(var/obj/machinery/power/supermatter_shard/S in supermatters)
+				if(S.UID() == newuid)
+					active = S
+
+		if("back")
+			active = null
+
```

Comment:
I'd suggest adding a default case with `else` that returns `FALSE` so that the UI doesn't refresh on an invalid action. Additionally `add_fingerprint` could be called after the switch to indicate that the computer was indeed used.